### ISSUE TITLE
Segmentation fault on Ubuntu 18.04

### ISSUE THREAD

**Issue by Janrupf** [NovaMods/nova-renderer#127]
This is the cleaned version of #126.
I fixed bugs that were caused due to Linux handling of enviroment variables, my Minecraft and VulkanSDK setup. I'm not good and not bad at C++, still learning, so if I did something wrong tell me. Anyway, the segmentation fault is caused at
https://github.com/NovaMods/nova-renderer/blob/7379f0339db151ec25d5c23ce8517e686753e639/src/main/cpp/render/objects/resources/shader_resource_manager.cpp#L209
which gives the following error messages:
```
[09:55:09] [Nova Renderer/ERROR]: ERROR: API: Validation Object: VK_NULL_HANDLE (Type = 0) | vkUpdateDescriptorSets: required parameter pDescriptorWrites[0].pImageInfo[0].imageView specified as VK_NULL_HANDLE
[09:55:09] [Nova Renderer/ERROR]: ERROR: API: Validation Object: VK_NULL_HANDLE (Type = 14) | Invalid ImageView Object 0x0. The spec valid usage text states 'If descriptorType is VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER, VK_DESCRIPTOR_TYPE_SAMPLED_IMAGE, VK_DESCRIPTOR_TYPE_STORAGE_IMAGE, or VK_DESCRIPTOR_TYPE_INPUT_ATTACHMENT, the imageView and imageLayout members of each element of pImageInfo must be a valid VkImageView and VkImageLayout, respectively' (https://www.khronos.org/registry/vulkan/specs/1.0/html/vkspec.html#VUID-VkWriteDescriptorSet-descriptorType-00326)
[09:55:09] [Nova Renderer/ERROR]: ERROR: API: Validation Object: 0x49 (Type = 23) | vkUpdateDescriptorSets() failed write update validation for Descriptor Set 0x49 with error: Write update to descriptor in set 0x49 binding #0 failed with error message: Attempted write update to combined image sampler descriptor failed due to: Invalid VkImageView: 0. The spec valid usage text states 'If descriptorType is VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER, VK_DESCRIPTOR_TYPE_SAMPLED_IMAGE, VK_DESCRIPTOR_TYPE_STORAGE_IMAGE, or VK_DESCRIPTOR_TYPE_INPUT_ATTACHMENT, the imageView and imageLayout members of each element of pImageInfo must be a valid VkImageView and VkImageLayout, respectively' (https://www.khronos.org/registry/vulkan/specs/1.0/html/vkspec.html#VUID-VkWriteDescriptorSet-descriptorType-00326)
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007fa3eee6c326, pid=31434, tid=0x00007fa466443700
#
# JRE version: Java(TM) SE Runtime Environment (8.0_171-b11) (build 1.8.0_171-b11)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.171-b11 mixed mode linux-amd64 compressed oops)
# Problematic frame:
# C  [libnvidia-glcore.so.390.48+0x13af326]
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /home/jannis/Dokumente/Coding/Minecraft/nova-renderer/jars/hs_err_pid31434.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#

```
Me as a learning C++ developer tried to debug this, maybe in a bad way. So I added the follwing code before `updateDescriptorSets`.
```c++
if(!writes[0].pImageInfo[0].imageView) {
    LOG(FATAL) << "Invalid image view at writes[0]";
}
```
Now it goes crazy... I do not get this error message and the segmentation fault does not occur any more, at least not here. Now it happens with that message:
```
[10:05:20] [Nova Renderer/INFO]: Creating descriptor sets for pipeline gui
[10:05:20] [Nova Renderer/DEBUG]: shader_resource_manager.cpp 79 - We've created 3 sets
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007ff1a70609ea, pid=32124, tid=0x00007ff218c9e700
#
# JRE version: Java(TM) SE Runtime Environment (8.0_171-b11) (build 1.8.0_171-b11)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.171-b11 mixed mode linux-amd64 compressed oops)
# Problematic frame:
# C  [libnova-renderer.so+0x8259ea]  nova::shader_resource_manager::update_all_descriptor_sets(nova::material_pass const&, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, nova::resource_binding, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, nova::resource_binding> > > const&)+0x486
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /home/jannis/Dokumente/Coding/Minecraft/nova-renderer/jars/hs_err_pid32124.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
```

I know you need the full logs, so here they are:
Without my modification: https://paste.ee/p/vdpf4
With my modification: https://paste.ee/p/f0BU8

**Comment by DethRaid** [comment]
Check out line 216 of your log. Looks like I reversed a texture format - which, strangely, causes no errors on my GTX 1080 on Windows 10.

Not sure if that's causing the error, but it's _an_ error

**Comment by Janrupf** [comment]
Maybe some drivers handling thibgs differently? But it might cause the error. Vulkan says that the imageView is a null handle, which is not true. Maybe something gets corrupted way down in Vulkan?

**Comment by Janrupf** [comment]
Ok, looks like this is going to be a long way :D
This is now segfaulting somewhere after destroing the render context... 

And yes, I made I stupid mistake...
```c++
if(!writes[0].pImageInfo[0].imageView) {
    LOG(FATAL) << "Invalid image view at writes[0]";
}
```
should be
```
if(!writes.at(0).pImageInfo->imageView) {
   LOG(FATAL) << "Invalid image view at writes[0]";
}
```

Ok, so now the new segfault:
```
(null)(ERROR / SPEC): msgNum: 629146858 - OBJ ERROR : For device 0x7f75e52476f0, SwapchainKHR object 0xe has not been destroyed. The spec valid usage text states 'All child objects created using instance must have been destroyed prior to destroying instance' (https://www.khronos.org/registry/vulkan/specs/1.0/html/vkspec.html#VUID-vkDestroyInstance-instance-00629)
    Objects: 1
       [0] 0xe, type: 1000001000, name: (null)
Validation(ERROR): msg_code: 629146858: Object: 0xe (Type = 27) | OBJ ERROR : For device 0x7f75e52476f0, SwapchainKHR object 0xe has not been destroyed. The spec valid usage text states 'All child objects created using instance must have been destroyed prior to destroying instance' (https://www.khronos.org/registry/vulkan/specs/1.0/html/vkspec.html#VUID-vkDestroyInstance-instance-00629)
[02:56:45] [Nova Renderer/TRACE]: render_context.cpp 280 - Destroyed the render context
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f756de88e20, pid=10409, tid=0x00007f75eb440700
#
# JRE version: Java(TM) SE Runtime Environment (8.0_171-b11) (build 1.8.0_171-b11)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.171-b11 mixed mode linux-amd64 compressed oops)
# Problematic frame:
# C  0x00007f756de88e20
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /home/jannis/Dokumente/Coding/Minecraft/nova-renderer/jars/hs_err_pid10409.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
#
```

And as a complete log: https://paste.ee/p/JPJyC

**Comment by Janrupf** [comment]
# Fix
GOT IT TO RUN!!!!!
I wonder how this even works on Windows...
So, here are the changes I made (It took me about 1 and a half hour to figure out what went wrong):
This:
https://github.com/NovaMods/nova-renderer/blob/1994c475515c12384b35713021e97c80dab9b714/src/main/cpp/render/objects/resources/shader_resource_manager.cpp#L193-L196
 needs to be changed to:
```c++
auto *image_info = new vk::DescriptorImageInfo();
image_info->setImageView(texture.get_image_view());
image_info->setImageLayout(vk::ImageLayout::eShaderReadOnlyOptimal);
image_info->setSampler(get_point_sampler());
```
and this:
https://github.com/NovaMods/nova-renderer/blob/1994c475515c12384b35713021e97c80dab9b714/src/main/cpp/render/objects/resources/shader_resource_manager.cpp#L198-L199
of course to this:
```c++
write.setPImageInfo(image_info)
    .setDescriptorType(vk::DescriptorType::eCombinedImageSampler);
```

## Explanation
What is happening here is that image_info gets out of scope and gets deleted. Maybe your compiler fixes that for you, but gcc doesn't.
**Please note that I did not deallocate this somewhere, so this might have to be done!**

# And another issue
Now I run into the next issue, or is just not implemented? Anyway, I just get a black screen.

**Comment by Janrupf** [comment]
Any Ideas, @DethRaid?

**Comment by DethRaid** [comment]
Yeah black screen is expected if you're not on the super latest commit. I only got things to be not black a couple days ago

For the other bit... Yeah, I'm not sure how that worked in the first place. I'm using GCC as well, on MinGW-w64 on Windows 10. I'll make the fix and get it in

**Comment by Janrupf** [comment]
Ok :D, gonna let git update my branch. I'll leave this open until every thing is working like expected on Linux. Btw, I would like to help you to get this project to forge (with SpongeMixins).

**Comment by Janrupf** [comment]
And you created an other reason to segfault somewhere :+1:. Gonna fix them all :laughing:

**Comment by DethRaid** [comment]
I'm surprised you loaded a world. Every time I try to do that it crashes

You're interested in porting Nova to forge? Join the discord server and ping me there (I have the same username as here). A couple others have done a little bit towards that goal and I can get y'all's in contact

**Comment by DethRaid** [comment]
Fixes for those issues are in the `vulkan` branch as of commit f6474a2, so I'm closing this ticket. I still definitely encourage you to join the Nova Discord (https://discord.gg/VGqtadw) so I can put you in touch with the other people interested in moving Nova to Forge
